### ISSUE TITLE
[Tracking] Remove all uses of `lib.mdDoc`

### ISSUE THREAD

**Issue by philiptaron** [NixOS/nixpkgs#300735]
Since https://github.com/NixOS/nixpkgs/pull/237557 was merged, `lib.mdDoc` has been a no-op. Let's remove the places that use it.

```
$ rg -w '(lib\.)?mdDoc' -o --no-filename | sort | uniq -c
  12356 lib.mdDoc
   1242 mdDoc
```

**Comment by hojerst** [comment]
as this seems to break the current home-manager project - wouldn't it be reasonable to wait a bit longer / issue a warning when this is used instead of removing it alltogether?

**Comment by eclairevoyant** [comment]
> wouldn't it be reasonable to wait a bit longer / issue a warning

There is a warning, unfortunately it's broken.

```
nix-repl> lib.options.mdDoc "test"
trace: warning: lib.mdDoc was removed from nixpkgs. Option descriptions are now in Markdown by default, you can remove any remaining uses of it.
"test"

nix-repl> lib.mdDoc "test"
error: attribute 'mdDoc' missing
       at «string»:1:1:
            1| lib.mdDoc "test"
             | ^
```

**Comment by philiptaron** [comment]
> as this seems to break the current home-manager project - wouldn't it be reasonable to wait a bit longer / issue a warning when this is used instead of removing it alltogether?

I'm mostly interested in removing the _use_ of `lib.mdDoc` in `nixpkgs` itself. Obviously there's a much longer period where it remains accessible from `lib`, likely originating from `lib/deprecated.nix`.

**Comment by philiptaron** [comment]
Closed by https://github.com/NixOS/nixpkgs/pull/303841 (thanks @stuebinm!)

**Comment by stuebinm** [comment]
ah sorry for the breakage everyone — I added that warning assuming it would be fine, but missed that `mdDoc` was only exposed via an `inherit` into `lib.mdDoc` (which of course got automatically removed along with all the other inherits).

**Comment by eclairevoyant** [comment]
No worries, it'll be fixed soon enough. Thanks for the original PR; maybe this will get HM to remove the usages sooner too ;)

**Comment by eclairevoyant** [comment]
Giving a reasonable deprecation period is how we should handle anything removed from `lib`. In this case it was removed from hm and nix-darwin fairly quickly, but apparently we still have [a lot more consumers using this function](https://github.com/search?q=mddoc+lang%3Anix&type=code). At the very least they should know why they are getting this error for now.

> Currently, `lib.mdDoc` still seems to exist but print a warning (see #304277) which consequently makes our builds fail

Besides, I do find it a little strange that simply _defining_ the function in nixpkgs _without using_ it would fail the CI.

**Comment by stuebinm** [comment]
@lorenzleutgeb you should only get this warning if the `lib.mdDoc` function is actually used somewhere within the code you're evaluating. Unfortunately, due to how `builtins.trace` works, there is no easy way to find out where that use site is, which can indeed be annoying if it's not in your own code but somewhere in your dependencies (indeed the definition of `lib.warn` in `lib/trivial.nix` [has a TODO item attached to it](https://github.com/NixOS/nixpkgs/blob/e8e9c5c3b1253e9347ff97a066a60746113cbfd8/lib/trivial.nix#L673-L674) that essentially says "figure out a way to show users where the warning came from").

but from quickly looking at your repo, a reasonable bet is that you're getting it since `dream2nix` [still has some uses of `mdDoc` in it](https://github.com/search?q=repo%3Anix-community%2Fdream2nix%20mdDoc&type=code), and you depend on it.

**Comment by lorenzleutgeb** [comment]
@stuebinm The offending option is actually printed: https://github.com/ngi-nix/ngipkgs/actions/runs/8780863139/job/24091654727#step:4:108 It seems that this occurrence of `lib.mdDoc` has been removed in the meantime, so I'll try updating the nixpkgs input of that Flake. We evaluate all of NixOS since we're generating Option documentation, so it's not really relevant to trace down single occurrences. If it's somewhere in NixOS, it will fail our build.

**Comment by eclairevoyant** [comment]
seems like it was fixed in #305379, so you'd need a commit after d9ad2e5a1a4ed119546f026aad1f167b1678b061
